cpp/ql/src/Likely Bugs/Underspecified Functions/TooFewArguments.qll

```diff
@@ -51,5 +51,7 @@ predicate tooFewArguments(FunctionCall fc, Function f) {
     hasDefiniteNumberOfParameters(fde)
   |
     fde.getNumberOfParameters() > fc.getNumberOfArguments()
-  )
+  ) and
+  // Don't report on implicit function declarations, as these are likely extraction errors.
+  not f.getADeclarationEntry().isImplicit()
```

Comment:
I wonder if this is too strict, given the test results that disappear. Shouldn't this be something like:
* if all declaration entries are explicit, then ignore
* if the non-explicit declaration entries differ on the number of parameters, then ignore

Reply:
One of the scenarios I'd like to handle is something like

file1.c:

```c
// Decl/definition
void assert(pchar __file, unsigned int __line, pchar __function);
```

file2.c:
```c
assert(x);  // Implicit fn decl because the assert macro isn't defined
```

In this case, some fndecl entries are explicit, and all explicit fndecls have the same number of parameters.

So I *can* write 

```ql
forex(FunctionDeclarationEntry fde | fde = f.getADeclarationEntry() |
    fde.getNumberOfParameters() = f.getNumberOfParameters()
  )
```

but this basically covers the case above. If we weaken it to

```ql
  forex(FunctionDeclarationEntry fde | fde = f.getAnExplicitDeclarationEntry() |
    fde.getNumberOfParameters() = f.getNumberOfParameters()
  ) and
```

then this doesn't handle the `assert` case above and doesn't get rid of all the noise.

Reply:
If this is just to handle `assert`: I would expect there to be some macro definition of `assert` to be in scope. Unless the particular project you're looking at hand-rolled the definition, and we're not managing to pick up on the correct header file?

Reply:
> What if you move `not f.getADeclarationEntry().isImplicit()` into `hasDefiniteNumberOfParameters`?

That didn't remove the FPs. I think it's because implicit `FunctionDeclarationEntry` always have 0 parameters anyway.

Reply:
> If this is just to handle `assert`: I would expect there to be some macro definition of `assert` to be in scope. Unless the particular project you're looking at hand-rolled the definition, and we're not managing to pick up on the correct header file?

The FPs have also appeared on projects that don't involve macros or `assert`. In the case of `assert`, most likely the wrong header file was included or couldn't be found.

It would be possible to remove some FPs simply by saying `and not exists(Macro m | m.getName() = f.getName())` but that's very specific and doesn't cover the cases where there isn't a macro involved.

In the general case, we'd hope to improve header file resolution, but we should expect this to go wrong at times.

Reply:
>  I think it's because implicit FunctionDeclarationEntry always have 0 parameters anyway.

Correct. An implicit function declarations `foo` always have the form `extern int foo()`.